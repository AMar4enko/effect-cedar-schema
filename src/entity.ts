import { AST, Schema } from "@effect/schema"
import { Class, Struct, TaggedClass as TaggedClassImpl } from "@effect/schema/Schema"
import { CedarNamespace } from "./annotations.js"
import { Effect, Equal, FiberRef, identity, Match, Record as R } from "effect"
import { UnknownException } from "effect/Cause"
import { SerializedIdentifier, FullSerializedType, SerializedType } from "./types.js"
import { IdentifierAnnotationId, isUndefinedKeyword } from "@effect/schema/AST"
import * as Hash from "effect/Hash"

type EntitySchema = Schema.Schema.All & { [EntityTypeId]: typeof EntityTypeId }

export const EntityTypeId: unique symbol = Symbol(`effect-cedar/EntityTypeId`)

export interface SerializedEntity extends Hash.Hash {
  identifier: SerializedIdentifier
  attributes: Record<string, FullSerializedType>
}

const EntitiesRef = FiberRef.unsafeMake(new Map<string, SerializedEntity>())

export type CompilerFunction = (v: unknown) => Effect.Effect<FullSerializedType, Error, never>

export const getEntities = FiberRef.get(EntitiesRef).pipe(
  Effect.map((map) => [...map.values()])
)

export const makeSerialisedEntity = function (identifier: SerializedIdentifier, attributes: Record<string, FullSerializedType>): SerializedEntity {
  const id = { ...identifier }
  const idHash = Hash.hash(`${identifier.entityType}${identifier.entityId}`)
  Hash.cached(id, idHash)
  const o = {
    identifier: id,
    attributes
  }

  Hash.cached(o, idHash)

  return o as any
}

export type FieldsWithParents< Fields extends Schema.Struct.Fields, MembersOf extends EntitySchema[] = never> = 
  [MembersOf] extends [never] 
    ? Fields & { id: typeof Schema.String }
    : Fields & { id: typeof Schema.String; parents: Schema.optional<Schema.Array$<Schema.Union<MembersOf>>> }


export interface TaggedClass<Self, Tag extends string, Fields extends Struct.Fields, Proto> extends
    Class<
      Self,
      Fields,
      Struct.Encoded<Fields>,
      Struct.Context<Fields>,
      Struct.Constructor<Omit<Fields, "_tag">>,
      {},
      Proto
    >
  {
    readonly _tag: Tag
  }

export interface Entity<Self, Tag extends string, Fields extends Schema.Struct.Fields, MembersOf extends EntitySchema[] = never> extends 
  TaggedClass<
    Self,
    Tag,
    FieldsWithParents<Fields, MembersOf>,
    {
      serialize: () => Effect.Effect<FullSerializedType, never, never>
    }
  >
  {
    [EntityTypeId]: typeof EntityTypeId
    readonly _tag: Tag
  }

/**
 * Gotta prohibit entity fields to be classes
 */
export type EntityField<F> = F extends new (...args: any[]) => any
  ? never
  : | Schema.Schema.All
    | Schema.PropertySignature.All

export const compile: (ast: AST.AST) => CompilerFunction = Match.type<AST.AST>().pipe(
  Match.when(
      (ast) => ast._tag === `Declaration` && ast.annotations[EntityTypeId] === EntityTypeId,
      () => (value: InstanceType<Entity<any, any, any, any>>) => value.serialize()
  ),
  Match.tag(`TupleType`, (ast) => {
    const runCompile = compile(ast.rest[0].type)

    return (value: any[]): Effect.Effect<SerializedType[], never, never> => Effect.all(value.map(runCompile)) as any
  }),
  Match.tag(`Union`, (ast): CompilerFunction => {
    const hasUndefined = ast.types.some((type) => isUndefinedKeyword(type))
    const withoutUndefined = ast.types.filter((type) => !isUndefinedKeyword(type))
    const map = new Map<string, CompilerFunction>()

    if (withoutUndefined.length === 1) {
      return compile(withoutUndefined[0])
    }

    return (value: unknown) => {
      if (value === undefined && hasUndefined) {
        return Effect.succeed(value)
      }

      /** Union here can only be union of multiple entity types, hence we are safe to assume we can serialize it */
      return ((value as any).serialize as any)()
    }
  }),
  Match.tag(`TypeLiteral`, (ast): any => {
    const compilers = R.fromEntries(
      ast.propertySignatures.map((propSignature) => [String(propSignature.name), compile(propSignature.type)] as const)
    )

    return (value: Record<PropertyKey, any>): Effect.Effect<FullSerializedType, never, never> => Effect.all(R.map(compilers, (runCompile, key) => runCompile(value[key]))) as any
  }),
  Match.orElse(() => Effect.succeed)
)
  
const toApiJSON = (schema: Entity<any, any, any, any>) => {
  const ast = schema.ast as AST.Transformation
   if (ast.to._tag !== `Declaration` || (ast.to.typeParameters.length === 0 || ast.to.typeParameters[0]._tag !== `TypeLiteral`)) { 
    throw new UnknownException(`unexpected AST`)
  } else {}
  const typeLiteral = ast.to.typeParameters[0]

  const NS = ast.to.annotations[CedarNamespace]
  const identifier: Record<`entityType` | `entityId`, (value: any) => string> = {
    entityType: () => `${NS}::${ast.to.annotations[IdentifierAnnotationId]}`,
    entityId: identity
  }
  const attributes: Record<string, CompilerFunction> = {}
  
  typeLiteral.propertySignatures.forEach((propSignature) => {
    if (propSignature.name === `id`) {
      identifier.entityId = (value: any) => value[`id`]
    } else if (propSignature.name !== `_tag` && !isUndefinedKeyword(propSignature.type)) {
      attributes[String(propSignature.name)] = compile(propSignature.type)
    }
  })
  
  return (value: Record<PropertyKey, any>): Effect.Effect<FullSerializedType, never, never> => Effect.gen(function* () {
    const a = yield* FiberRef.get(EntitiesRef)

    const key = `${identifier.entityType(value)}_${identifier.entityId(value)}`

    return yield* Effect.fromNullable(a.get(key))
      .pipe(
        Effect.orElse(() => 
          Effect.all(R.map(attributes, (runCompile, key) => runCompile(value[key]))).pipe(
            Effect.map((attributes) => 
              makeSerialisedEntity(
                R.map(identifier, (a) => a(value)),
                attributes
              )
            )
          ).pipe(
            Effect.tap((value) => FiberRef.update(EntitiesRef, (a) => a.set(key, value)))
          )
        ),
        Effect.map(({ identifier }) => identifier)
      )
  }) as any
}

export const Entity
  = <Self>() =>
  <Tag extends string, Fields extends Schema.Struct.Fields, MembersOf extends EntitySchema[] = never>(
    tag: Tag, 
    args: {
      fields: Fields,
      membersOf?: MembersOf
    },
    namespace: string
  ): Entity<Self, Tag, Fields, MembersOf> => {
      const membersOf = Schema.optional(args.membersOf ? Schema.Array(Schema.Union(...args.membersOf)) : Schema.Undefined)
      const annotations = namespace ? { [CedarNamespace]: namespace } : undefined
      const base = (TaggedClassImpl<Self>(tag)(tag, { ...args.fields, id: Schema.String, parents: membersOf }, { ...annotations, [EntityTypeId]: EntityTypeId }) as any)

      const serialize = toApiJSON(base)

      class Class extends base {
        [EntityTypeId] = EntityTypeId
        namespace = namespace

        serialize() {
          return serialize(this)
        }
      }

      return Class as any
    }