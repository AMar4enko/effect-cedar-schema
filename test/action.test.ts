import { describe, expect, it } from 'vitest'
import { CancelOrder, PlaceOrder } from './fixtures/actions.js'
import * as ent from './fixtures/entities.js'
import { Effect, Hash } from 'effect'
import { getEntities, makeSerialisedEntity } from '../src/entity.js'
import { equals } from 'effect/Equal'

describe(`Action`, () => {
  it(`should serialize`, () => {
    const managerRole = new ent.Role({
      id: `managers`,
      name: `Managers`
    })

    const manager = new ent.User({
      id: `1`,
      name: `user`,
      parents: [
        managerRole
      ]
    })

    const order = new ent.Order({
      id: `2`,
      manager,
      stock: {
        available: 10,
        reserved: 0
      },
      storage: new ent.Storage({
        id: `1`,
        location: `storage`
      }),
      books: [
        new ent.Book({
          id: `1`,
          title: `book`,
          ageFrom: 10,
          author: new ent.Author({
            id: `1`,
            name: `author`
          }),
        })
      ]
    })

      
    const placeOrder = new PlaceOrder({
      resource: order
    })

    const cancelOrder = new CancelOrder({
      principal: manager,
      resource: order,
      context: {
        reason: `Cancelled by customer`
      }
    })

    const actions = Effect.all([placeOrder.serialize(), cancelOrder.serialize()])

    const serialize = actions.pipe(
      Effect.bindTo(`actions`),
      Effect.bind(`entities`, () => getEntities)  
    )
    
    const result = Effect.runSync(serialize)

    expect(result).toMatchSnapshot()
  })
})