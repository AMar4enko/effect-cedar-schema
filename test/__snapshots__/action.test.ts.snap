// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Action > should serialize 1`] = `
{
  "actions": [
    {
      "action": {
        "actionId": "placeOrder",
        "actionType": "BookStore::Action",
      },
      "resource": {
        "entityId": "2",
        "entityType": "BookStore::Order",
      },
    },
    {
      "action": {
        "actionId": "cancelOrder",
        "actionType": "BookStore::Action",
      },
      "context": {
        "contextMap": {
          "author": {
            "entityIdentifier": {
              "entityId": "1",
              "entityType": "BookStore::Author",
            },
          },
          "reason": {
            "string": "Cancelled by customer",
          },
        },
      },
      "principal": {
        "entityId": "1",
        "entityType": "BookStore::User",
      },
      "resource": {
        "entityId": "2",
        "entityType": "BookStore::Order",
      },
    },
  ],
  "entities": [
    {
      "attributes": {
        "name": {
          "string": "author",
        },
      },
      "identifier": {
        "entityId": "1",
        "entityType": "BookStore::Author",
      },
      "parents": [],
    },
    {
      "attributes": {
        "ageFrom": {
          "long": 10,
        },
        "author": {
          "entityIdentifier": {
            "entityId": "1",
            "entityType": "BookStore::Author",
          },
        },
        "title": {
          "string": "book",
        },
      },
      "identifier": {
        "entityId": "1",
        "entityType": "BookStore::Book",
      },
      "parents": [],
    },
    {
      "attributes": {
        "location": {
          "string": "storage",
        },
      },
      "identifier": {
        "entityId": "1",
        "entityType": "BookStore::Storage",
      },
      "parents": [],
    },
    {
      "attributes": {
        "name": {
          "string": "Managers",
        },
      },
      "identifier": {
        "entityId": "managers",
        "entityType": "BookStore::Role",
      },
      "parents": [],
    },
    {
      "attributes": {
        "name": {
          "string": "user",
        },
      },
      "identifier": {
        "entityId": "1",
        "entityType": "BookStore::User",
      },
      "parents": [
        {
          "entityId": "managers",
          "entityType": "BookStore::Role",
        },
      ],
    },
    {
      "attributes": {
        "books": {
          "set": [
            {
              "entityIdentifier": {
                "entityId": "1",
                "entityType": "BookStore::Book",
              },
            },
          ],
        },
        "manager": {
          "entityIdentifier": {
            "entityId": "1",
            "entityType": "BookStore::User",
          },
        },
        "stock": {
          "record": {
            "available": {
              "long": 10,
            },
            "reserved": {
              "long": 0,
            },
          },
        },
        "storage": {
          "entityIdentifier": {
            "entityId": "1",
            "entityType": "BookStore::Storage",
          },
        },
      },
      "identifier": {
        "entityId": "2",
        "entityType": "BookStore::Order",
      },
      "parents": [],
    },
  ],
}
`;
